### PR TITLE
Weaken `restoreFileFromSnapshot` assertion

### LINKED ISSUE #96018: [CI] RestoreSnapshotIT testChangeSettingsOnRestore failing
**Build scan:**
https://gradle-enterprise.elastic.co/s/57brnuf4tdwpq/tests/:server:internalClusterTest/org.elasticsearch.snapshots.RestoreSnapshotIT/testChangeSettingsOnRestore

**Reproduction line:**
```
./gradlew ':server:internalClusterTest' --tests "org.elasticsearch.snapshots.RestoreSnapshotIT.testChangeSettingsOnRestore" -Dtests.seed=E04845BEA9B6442F -Dtests.locale=uk -Dtests.timezone=America/Anguilla -Druntime.java=20
```

**Applicable branches:**
7.17

**Reproduces locally?:**
No

**Failure history:**
https://gradle-enterprise.elastic.co/scans/tests?tests.container=org.elasticsearch.snapshots.RestoreSnapshotIT&tests.test=testChangeSettingsOnRestore

**Failure excerpt:**
```
java.lang.AssertionError: All incoming requests on node [node_s3] should have finished. Expected 0 but got 171; pending tasks [[]]

  at org.elasticsearch.test.InternalTestCluster.lambda$assertRequestsFinished$43(InternalTestCluster.java:2786)
  at org.elasticsearch.test.ESTestCase.assertBusy(ESTestCase.java:1141)
  at org.elasticsearch.test.InternalTestCluster.assertRequestsFinished(InternalTestCluster.java:2775)
  at org.elasticsearch.test.InternalTestCluster.assertAfterTest(InternalTestCluster.java:2749)
  at org.elasticsearch.test.ESIntegTestCase.afterInternal(ESIntegTestCase.java:622)
  at org.elasticsearch.test.ESIntegTestCase.cleanUpCluster(ESIntegTestCase.java:2396)
  at jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
  at java.lang.reflect.Method.invoke(Method.java:578)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1758)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:1004)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:49)
  at org.apache.lucene.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:45)
  at org.apache.lucene.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:48)
  at org.apache.lucene.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:64)
  at org.apache.lucene.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:47)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.forkTimeoutingTask(ThreadLeakControl.java:843)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$3.evaluate(ThreadLeakControl.java:490)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.runSingleTest(RandomizedRunner.java:955)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$5.evaluate(RandomizedRunner.java:840)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:891)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:902)
  at org.apache.lucene.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:45)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.util.TestRuleStoreClassName$1.evaluate(TestRuleStoreClassName.java:41)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53)
  at org.apache.lucene.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:47)
  at org.apache.lucene.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:64)
  at org.apache.lucene.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:54)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.lambda$forkTimeoutingTask$0(ThreadLeakControl.java:850)
  at java.lang.Thread.run(Thread.java:1623)

```

### PR DESCRIPTION
**Pull request by DaveCTurner**
It's possible to call `RecoveryTarget#restoreFileFromSnapshot` after the
`RecoveryTarget` has been replaced with a new instance due to a retry,
but before all its refs have been released. If the recovery was
snapshot-based then in this situation we will have already transferred
the permits to the new instance, so the assertion that this instance has
permits will trip. This commit fixes the problem with a non-null
placeholder value indicating that the recovery target was at least
originally created while holding some snapshot recovery permits.

Closes #96018

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/indices/recovery/RecoveryTarget.java [hunk 1]
```diff
@@ -153,7 +156,7 @@ public RecoveryTarget retryCopy() {
         // If we're retrying we should remove the reference from this instance as the underlying resources
         // get released after the retry copy is created
         Releasable snapshotFileDownloadsPermitCopy = snapshotFileDownloadsPermit;
-        snapshotFileDownloadsPermit = null;
+        snapshotFileDownloadsPermit = SNAPSHOT_FILE_DOWNLOADS_PERMIT_PLACEHOLDER_FOR_RETRY;
```
Comment (kingherc) [inline server/src/main/java/org/elasticsearch/indices/recovery/RecoveryTarget.java:159]: > This commit fixes the problem with a non-null placeholder value indicating that the recovery target was at least originally created while holding some snapshot recovery permits.

Should we add an assertion that snapshotFileDownloadsPermit != null before replacing its value? To assert that the original was holding some snapshot recovery permits as per your comment.

File: server/src/main/java/org/elasticsearch/indices/recovery/RecoveryTarget.java [hunk 2]
```diff
@@ -585,6 +588,7 @@ public void restoreFileFromSnapshot(
         BlobStoreIndexShardSnapshot.FileInfo fileInfo,
         ActionListener<Void> listener
     ) {
+        assert hasReferences();
```
Comment (kingherc) [inline server/src/main/java/org/elasticsearch/indices/recovery/RecoveryTarget.java:591]: This is new, I guess orthogonal to the PR, but still valid in all cases?

File: server/src/main/java/org/elasticsearch/indices/recovery/RecoveryTarget.java (hunk 1 above)
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/indices/recovery/RecoveryTarget.java]: Fair point. I don't think it really matters but I made sure we don't go from `null` to non-`null` here in 91ec6cb6a6a8.

File: server/src/main/java/org/elasticsearch/indices/recovery/RecoveryTarget.java (hunk 2 above)
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/indices/recovery/RecoveryTarget.java:593]: Yep

**Review by henningandersen — APPROVED** [review]
LGTM.

**Review by kingherc — APPROVED** [review]
Thanks for the explanations and changes! LGTM

**Pull request merged**